proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,199 @@
+once the system setting is in dark mode, thereby enhancing the overall browser
+experience.
+
+#### Use Cases
+
+1. Improved icon visibility in the extension toolbar.
+1. Improved icon visibility on management the management and shortcuts pages.
+1. Dark mode icon declarations made possible through the extension manifest.
+1. setIcon() will allow setting of dark and/or light mode specific icons.
+1. Improved icon visibility on context menu (Chrome and Safari reply on default
+icons, only Firefox can specify icons).
+1. Improved icon visibility on side panel (Chrome replies on default icons,
+Firefox can specify icons).
+
+### Known Consumers
+
+The Chromium bug has a significant amount of developer interest.
+
+## Specification
+
+### Schema
+
+manifest.json
+```
+"icon_variants": [
+  {
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png",
+  },
+  {
+    "16": "dark16.png",
+    "32": "dark32.png",
+    "color_scheme": "dark",
+  },
+  {
+    "16": "light16.png",
+    "32": "light32.png",
+    "color_scheme": ["dark", "light"]
+  }
+]
+```
+
+Set icon_variants dynamically.
+```
+const exampleProperties: {string: string | ImageData}[] = [
+  {
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png"
+  },
+  {
+    "16": darkImageData16,
+    "32": darkImageData32,
+    "color_scheme": "dark"
+  },
+  {
+    "16": lightImageData16,
+    "32": lightImageData32,
+    "color_scheme": "light"
+  }
+];
+
+// action.setIcon().
+const actionProperties = {variants: exampleProperties};
+action.setIcon(createProperties);
+
+// menus.*(), for supporting browsers.
+const menusProperties = {icon_variants: exampleProperties}
+menus.create(menusProperties);
+menus.update(menusProperties);
+```
+
+A benefit of this new structure is that it's more resiliant to future changes, thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
+monochrome), and etc.
+
+## setIcon()
+Incumbent action.setIcon(), for reference.
+```
+action.setIcon({
+  path,
+  tabId
+  imageData,
+});
+```
+
+### Behavior
+
+Existing manifest key. The behavior of the icons manifest key will remain unchanged.
+```
+{
+  "icons": {
+    "64": "icon_64.png"
+  },
+  "action": {
+    "default_icon": {
+      "64": "action_64.png"
+    }
+  }
+}
+```
+
+Order of precedence. The new `icon_variants` keys and subkeys will take
+precedent, followed by the incumbent `icons` key.
+
+The `dark` icon variant is used when a dark color scheme is used for the
+surrounding UI, often referred to as "dark mode".
+The `light` icon variant is used otherwise.
+
+### New Permissions
+
+N/A.
+
+### Manifest File Changes
+
+1. If icon_variants are supplied, icons can be ignored.
+1. If icon_variants contain an icon object duplicate, the first match will be
+used.
+1. Any icon object that does not contain a "color_scheme" key will apply to both
+light and dark.
+1. If only one icon object is supplied, it will be used for both light and dark.
```

Comment:
Even if `color_scheme: "dark"` is specified? I would lean towards saying that if the browser recognises the icon_variants key, it can error if there is not at least one light mode (or light and dark mode etc.) icon provided.

Reply:
@oliverdunk having a check if at least one light mode icon is specified would be helpful. If we make it a warning instead of an error we would need to figure out what to do instead. @hanguokai do you believe the browser should simply use the dark icon instead?

Reply:
@carlosjeurissen For now, browsers are quite tolerant of `icons`. For example, if developers don't provide `icons` or `icons` is an empty array (no one icon), the browser just ignore it and display a default icon without any error or warning. Of course, developers generally do not do this. However, what is discussed here is how browsers handle the most extreme situations. I think tolerating is better than reporting errors, just like now. And in my option, if only one icon is provided, the browser no longer needs to match any conditions. This is also a bit like the relationship between programming language engine and linter. The engine only needs to make sure it works properly, and the extra linter tool (e.g. a IDE addon) makes sure your code best.